### ISSUE TITLE
TeslaZap projectile overlap bug and WithMuzzleOverlay looking stupid

### ISSUE THREAD

**Issue by djohe** [OpenRA/OpenRA#16943]
This bug was dismissed in 2017 here but I do not agree with what obrakmann said about it the slightest: https://github.com/OpenRA/OpenRA/issues/13420
![OpenRA_Teslazap_proj_bug](https://user-images.githubusercontent.com/771748/62943732-6fa3bd00-bddb-11e9-92c7-752bc0526fcb.png)
WithMuzzleOverlay also looks really stupid for C&C1 flamethrowers & chemspray weaponry or any other weapon with long muzzle flash animation because you can move the units / infantry before the muzzle animation finishes and the muzzle follows the unit. WW had muzzleflash tied to weapons rather than the units/structures and it worked much better.
A possible solution could be to have a line that tells WithMuzzleOverlay to not follow the unit/structure after first tick or add muzzle flash to weaponry directly like WW games. Issue similar to this is: https://github.com/OpenRA/OpenRA/issues/10763

I also wish for a line to control the frame duration of a single TeslaZap projectile that is set to Duration: 1 (basically have it only render one shape for weapon firing once but have a shape duration of whatever I want ticks)

**Comment by GraionDilach** [comment]
While WW muzzleflashes were tied to the weapon, they were still following the weapon offset the way OpenRA does.

The sole exception to this is Tesla electric bolts fired by vehicles in Red Alert 2. Which sounds like a bug in the implementation, actually, since even Battle Fortresses filled with Tesla Troopers were working in YR.

**Comment by djohe** [comment]
Sorry I should have investigated better, you are correct in that muzzle flash follows the unit in WW games, however the muzzle flash does not change direction if you order a C&C1 flamethrower infantry to shoot to the right of him and then quickly order him to run to the left of him.

In WW games, atleast in C&C1 & TibSun & RA2:YR infantry does not face their target untill next time it shoots so if you order a flamethrower to first fire to the right of him and then to the left of him he wont immediately face left when you order, only faces left the next time he shoots.

Basically WW infantry do 360NOSCOPE but C&C1 the vehicles turn to face their target ASAP, in TibSun/RA2:YR non turreted vehicles only turn when they have reloaded their weapon (stupid). OpenRA infantry turn ASAP to face their target even when not reloaded weapon breaking the C&C1 flamethrower illusion.

**Comment by GraionDilach** [comment]
You can just pause `Mobile` with a `GrantConditionOnAttack` to achieve something similar to the stop-turning/moving effect as a modder. You don't need new logics for that one. And the GCoA frames provide better timing to this than a hardcoded weapon delay, allowing you to let the unit turning immediately after the muzzle disappears.

Doing such a change as a the GCoA-Mobile relation on the default mods would have balance implications though, so that's a LETSNOTGOTHERE.

**Comment by djohe** [comment]
Yes yes I know this is OpenRA and not orignal C&C / RA1 / whatever game you try to recreate in OpenRA and GrantConditionOnAttack is a fugly hack for C&C1 flamethrower infantry that is visually broken and GCoA-Mobile makes them respond worse / balance implications just like you said yourself.
It is as fugly as putting a dummy weapon on C&C1 MSAM (Rocket Launcher) just to get NoMovingFires=yes to work on it and even then it does not work good. (Sorry about my tone in these issue posts, I am not mad, just kinda disappointed in how some OpenRA stuff works, alot is good tho)

Hell I even just remembered another issue related to infantry & firing animation: the RA1 mods flamethrower infantry having its multi burst flamethrower causing his animation to reset each burst and does not show him visually lit up untill last burst.

**Comment by GraionDilach** [comment]
`NoMovingFire` is also covered with GCoA-Mobile interaction - actually that is the setup I suggested, since turn is a move. The infantry burst issue is something I haven't bothered with enough yet to mod around with a solution.

I wouldn't call it a fugly hack though. pausing would still allow orders be received and executed - there would be no orders dropped using this idea. It would just give the game a different feel which probably be considered as a step-back to the simplicity arcadish forwardness of what the competitive community prefers.

**Comment by reaperrr** [comment]
In my opinion, ZOffset not having the desired effect even in mods that use neither isometric perspective nor DepthBuffer can be considered a bug.
